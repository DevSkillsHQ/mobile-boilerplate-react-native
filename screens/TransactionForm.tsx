import React, { useState } from 'react'
import { StyleSheet, TextInput, Button, Keyboard } from 'react-native'
import { Text, View } from '../components/Themed'
import { BASE_API_URL } from '../env'
import { useAccountContext } from '../context/AccountContext'

export const TransactionForm = () => {
  const [text, setText] = useState<string>('')
  const [hasError, setHasError] = useState<string | null>(null)

  const account = useAccountContext()

  const handleClick = async () => {
    if (!account?.account_id) return

    const data = {
      account_id: account?.account_id,
      amount: text,
    }

    try {
      const response = await fetch(`${BASE_API_URL}/transactions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(data),
      })

      if (response.ok) {
        const responseData = await response.json()
        console.log('Transaction successful:', responseData)
      } else {
        setHasError(response.status.toString())
        console.error('Error adding transaction:', response.status)
      }
    } catch (error) {
      setHasError('Server error')
      console.error('Error adding transaction:', error)
    } finally {
      Keyboard.dismiss()
      setText('')
      setHasError(null)
    }
  }

  return (
    <View style={styles.container}>
      <Text>Form</Text>
      <TextInput
        keyboardType="numeric"
        style={styles.input}
        value={text}
        onChangeText={setText}
      />
      <Button onPress={handleClick} title="Add transaction" />
      {hasError !== null ? (
        <View style={styles.error}>
          <Text>Something went wrong with your transaction: {hasError}</Text>
        </View>
      ) : (
        <></>
      )}
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  separator: {
    marginVertical: 30,
    height: 1,
    width: '80%',
  },
  input: {
    width: '80%',
    height: 40,
    margin: 12,
    borderWidth: 1,
    padding: 10,
  },
  error: {
    margin: 10,
    backgroundColor: '#FFCCCC',
    padding: 5,
  },
})
