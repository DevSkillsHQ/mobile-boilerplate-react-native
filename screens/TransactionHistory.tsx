import React, { useCallback } from 'react'
import { StyleSheet, ScrollView, Button } from 'react-native'
import { Text, View } from '../components/Themed'
import { useFetch } from '../hooks/useFetch'
import { BASE_API_URL } from '../env'
import { useFocusEffect } from '@react-navigation/native'
import { TransactionItem } from '../components/TransactionItem'
import { Transaction } from '../types'

export const TransactionHistory = () => {
  const { error, data, refetch } = useFetch<Transaction[]>(
    `${BASE_API_URL}/transactions`
  )

  useFocusEffect(
    useCallback(() => {
      refetch()
    }, [])
  )

  if (error)
    return (
      <View style={styles.container}>
        <Button onPress={refetch} title="Refetch data" />
        <Text>
          Something went wrong in loading the content. Please refresh.
        </Text>
      </View>
    )

  return (
    <View style={styles.container}>
      <ScrollView>
        <Button onPress={refetch} title="Refetch data" />
        {data?.map((item: Transaction) => (
          <TransactionItem item={item} />
        ))}
      </ScrollView>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
  },
})
