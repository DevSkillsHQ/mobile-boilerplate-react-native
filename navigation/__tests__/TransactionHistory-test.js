import React from 'react'
import renderer, { act } from 'react-test-renderer'
import { TransactionHistory } from '../../screens/TransactionHistory'

jest.mock('../../hooks/useFetch', () => ({
  useFetch: jest.fn(() => ({
    loading: false,
    error: null,
    data: [
      {
        account_id: '0afd02d3-6c59-46e7-b7bc-893c5e0b7ac2',
        amount: 50,
        created_at: new Date(),
        transaction_id: 'abc123',
      },
    ],
    refetch: jest.fn(),
  })),
}))

jest.mock('../../context/AccountContext', () => ({
  useAccountContext: jest.fn(() => ({
    account_id: '0afd02d3-6c59-46e7-b7bc-893c5e0b7ac2',
    balance: 100,
  })),
}))

describe('navigation', () => {
  describe('TransactionHistory component', () => {
    test('it renders error state correctly', async () => {
      jest.spyOn(console, 'error').mockImplementation(() => {})
      jest
        .requireMock('../../hooks/useFetch')
        .useFetch.mockImplementation(() => ({
          loading: false,
          error: 'An error occurred',
          data: null,
          refetch: jest.fn(),
        }))

      let tree
      await act(async () => {
        tree = renderer.create(<TransactionHistory />)
        await new Promise((resolve) => setTimeout(resolve, 0))
      })
      expect(tree.toJSON()).toMatchSnapshot()
    })

    test('it renders data correctly', async () => {
      let tree
      await act(async () => {
        tree = renderer.create(<TransactionHistory />)
        await new Promise((resolve) => setTimeout(resolve, 0))
      })
      expect(tree.toJSON()).toMatchSnapshot()
    })
  })
})
