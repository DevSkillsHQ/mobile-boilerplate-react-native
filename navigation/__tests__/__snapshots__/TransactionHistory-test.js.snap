// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigation TransactionHistory component it renders data correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Refetch data
      </Text>
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#000",
        },
        undefined,
      ]
    }
  >
    Something went wrong in loading the content. Please refresh.
  </Text>
</View>
`;

exports[`navigation TransactionHistory component it renders error state correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Refetch data
      </Text>
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#000",
        },
        undefined,
      ]
    }
  >
    Something went wrong in loading the content. Please refresh.
  </Text>
</View>
`;

exports[`navigation TransactionHistory component it renders loading state correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <RCTScrollView>
    <View>
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Refetch data
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "backgroundColor": "papayawhip",
              "marginTop": 5,
              "padding": 15,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          Transferred $
          50
           from account 
          0afd02d3-6c59-46e7-b7bc-893c5e0b7ac2
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000",
              },
              undefined,
            ]
          }
        >
          The current balance is 
          100
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
